### PR TITLE
Use CAM service account instead of CAC token (multi-region)

### PR DESCRIPTION
**Pull request by Edwin-Pau**

### CODE REVIEW

File: deployments/gcp/multi-region/terraform.tfvars.sample [hunk 1]
```diff
@@ -72,7 +72,7 @@ centos_admin_ssh_pub_key_file = "~/.ssh/id_rsa.pub"
 # secrets. The GCP service account specified above in "gcp_service_account"
 # must be in the same project and have KMS decryptor permissions for this key.
 
-kms_cryptokey_id = "projects/<project-id>/locations/<location>/keyRings/<keyring-name>/cryptoKeys/<key-name>"
+# kms_cryptokey_id   = "projects/<project-id>/locations/<location>/keyRings/<keyring-name>/cryptoKeys/<key-name>"
```
Comment (shermanyin) [inline deployments/gcp/multi-region/terraform.tfvars.sample:75]: why the extra spaces before "="?

File: modules/gcp/cac-igm/main.tf [hunk 2]
```diff
@@ -35,6 +38,31 @@ data "google_compute_image" "cac-base-img" {
...
+  source  = var.cam_credentials_file
+}
+
+resource "google_storage_bucket_object" "cac-cam-script" {
+  count = local.num_cacs == 0 ? 0 : 1
+
+  bucket  = var.bucket_name
+  name    = local.cam_script_name
+  content = templatefile(
+    "${path.module}/${local.cam_script_name}.tmpl",
```
Comment (shermanyin) [inline modules/gcp/cac-igm/main.tf:55]: ideally we have a generic script, and these values are passed to the startup scripts which calls cac-cam.py with these values as command line arguments.

File: modules/gcp/cac-igm/main.tf [hunk 3]
```diff
@@ -35,6 +38,31 @@ data "google_compute_image" "cac-base-img" {
...
+    {
+      kms_cryptokey_id     = var.kms_cryptokey_id,
+      cam_credentials_name = local.cam_credentials_name,
+
+      ssl_key     = "",
```
Comment (shermanyin) [inline modules/gcp/cac-igm/main.tf:60]: why does the script need ssl_key/cert?

File: modules/gcp/cac-igm/main.tf [hunk 4]
```diff
@@ -7,8 +7,11 @@
 
 locals {
   prefix = var.prefix != "" ? "${var.prefix}-" : ""
-  startup_script = "cac-startup.sh"
-  num_cacs = length(flatten([for i in var.instance_count_list: range(i)]))
+  startup_script       = "cac-startup.sh"
+  cam_script_name      = "cac-cam.py"
+  cam_credentials_name = "cam-cred.json"
```
Comment (shermanyin) [inline modules/gcp/cac-igm/main.tf:12]: i'd call this cam_credentials_json or cam_credentials_file

File: modules/gcp/cac-igm/main.tf [hunk 5]
```diff
@@ -7,8 +7,11 @@
 
 locals {
   prefix = var.prefix != "" ? "${var.prefix}-" : ""
-  startup_script = "cac-startup.sh"
-  num_cacs = length(flatten([for i in var.instance_count_list: range(i)]))
+  startup_script       = "cac-startup.sh"
+  cam_script_name      = "cac-cam.py"
```
Comment (shermanyin) [inline modules/gcp/cac-igm/main.tf:11]: cam_script?

File: modules/gcp/cac-igm/cac-startup.sh.tmpl [hunk 6]
```diff
@@ -18,13 +18,32 @@ log() {
     echo "[$(date)] $${message}" | tee -a "$INSTALL_LOG"
 }
 
+get_cac_token() {
+    echo '### Retrieve connector token before CAC install ###'
+    log "Retrieving connector token..."
```
Comment (shermanyin) [inline modules/gcp/cac-igm/cac-startup.sh.tmpl:23]: maybe for another commit but we should clean up the use of log vs echo in this file

File: modules/gcp/cac-igm/cac-startup.sh.tmpl [hunk 7]
```diff
@@ -18,13 +18,32 @@ log() {
...
+
+    # Download and install python3
+    apt-get -qq update
+    apt install -y python3
```
Comment (shermanyin) [inline modules/gcp/cac-igm/cac-startup.sh.tmpl:27]: i think we should combine this and other stuff we need to install to one place where it install all requirements.  doing apt-get update a 2nd time is not needed.

File: modules/gcp/cac-igm/cac-startup.sh.tmpl [hunk 8]
```diff
@@ -44,13 +63,19 @@ get_credentials() {
...
-        data=$(echo "{ \"ciphertext\": \"${cac_token}\" }")
+        # Gets decrypted CAM service account credentials and outputs to plaintext json file
+        cam_cred_encrypted=$(cat ${cam_credentials_name})
+        data=$(echo "{ \"ciphertext\": \"$cam_cred_encrypted\" }")
         b64_data=$(curl -X POST -d "$data" "$DECRYPT_URI" -H "Authorization: Bearer $token" -H "Content-type: application/json" | python -c "import sys, json; print json.load(sys.stdin)['plaintext']")
-        CAC_TOKEN=$(echo "$b64_data" | base64 --decode)
+        CAM_CREDENTIALS=$(echo "$b64_data" | base64 --decode)
+        echo $CAM_CREDENTIALS | tee ${cam_credentials_name}
+
+        get_cac_token
```
Comment (shermanyin) [inline modules/gcp/cac-igm/cac-startup.sh.tmpl:73]: if you're running the same function at the end of both "if" and "else", put it outside the if/else clause

File: modules/gcp/cac-igm/cac-startup.sh.tmpl [hunk 9]
```diff
@@ -18,13 +18,32 @@ log() {
...
+
+    # Download and install python3
+    apt-get -qq update
+    apt install -y python3
+
+    # Download the CAM python script from the bucket and run it to create a cac token
+    gsutil cp gs://${bucket_name}/${cam_script_name} $INSTALL_DIR
+
+    # Set CAC_TOKEN variable using the script's output
+    CAC_TOKEN=`python3 ${cam_script_name}`
```
Comment (shermanyin) [inline modules/gcp/cac-igm/cac-startup.sh.tmpl:33]: your script already has "#!/usr/local/bin/python3" so I don't think you need to put python3 in front.

File: modules/gcp/cac-igm/cac-cam.py.tmpl [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+import datetime
+import json
+import os
+import requests
+import sys
+
+# Declare global variables
+API_URL              = "https://cam.teradici.com/api/v1"
+CAM_CREDENTIALS_FILE = "cam-cred.json"
```
Comment (shermanyin) [inline modules/gcp/cac-igm/cac-cam.py.tmpl:16]: API_URL should have cam.teradici.com as default but also allow over-ride with command line options
CAM_CREDENTAILS_FILE should be a command line argument.

File: modules/gcp/cac-igm/cac-cam.py.tmpl [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+# Set up connector name using the current time
+iso_time = datetime.datetime.utcnow().isoformat(timespec='seconds').replace(':','').replace('-','') + 'Z'
```
Comment (shermanyin) [inline modules/gcp/cac-igm/cac-cam.py.tmpl:19]: With many connectors, I think there will need to be a way for users to see which connector corresponds to which VM in the cloud.  So I think the hostname should be in here somewhere.  Also, we might want to also specify the region and/or zone that this VM is in so we know how many connectors are in each zones.  Maybe something like "us-west2-vm-cac-2"  or something like that? You should be able to find a bunch of relevant info from the metadata server.

File: modules/gcp/cac-igm/cac-cam.py.tmpl [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+    response_body   = response.json()
+    connector_token = response_body.get('data').get('token')
+
+    return connector_token
+
+
+# Use helper functions to return a cac_token string using CAM credentials file
+def main():
+    auth_token         = get_auth_token(CAM_CREDENTIALS_FILE)
+    cac_deployment_id  = get_deployment_id(auth_token)
```
Comment (shermanyin) [inline modules/gcp/cac-igm/cac-cam.py.tmpl:84]: Why is this needed? can't you get deploymentId from the CAM credentials file?

File: modules/gcp/cac-igm/cac-cam.py.tmpl [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    cac_token          = get_cac_token(auth_token, cac_deployment_id)
+
+    return cac_token
+
+
+# Entry point to this script
```
Comment (shermanyin) [inline modules/gcp/cac-igm/cac-cam.py.tmpl:90]: no need for this comment

File: modules/gcp/cac-igm/cac-cam.py.tmpl [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+if __name__ == '__main__':
+    try:
+        # Return the cac_token string as the output of this script
+        sys.stdout.write(main())
```
Comment (shermanyin) [inline modules/gcp/cac-igm/cac-cam.py.tmpl:94]: can you just use print()?

File: modules/gcp/cac-igm/cac-cam.py.tmpl [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+    except:
+        # Prevent bash from interpreting any error messages
+        sys.exit(1)
```
Comment (shermanyin) [inline modules/gcp/cac-igm/cac-cam.py.tmpl:97]: newline at end of file

File: modules/gcp/cac-igm/cac-startup.sh.tmpl (hunk 9 above)
Comment (shermanyin) [inline modules/gcp/cac-igm/cac-startup.sh.tmpl:33]: You should also check "$?" to see the previous command failed.

**Review by shermanyin — CHANGES_REQUESTED** [review]

File: modules/gcp/cac-igm/cac-cam.py.tmpl [hunk 16]
```diff
@@ -0,0 +1,97 @@
+# Copyright (c) 2020 Teradici Corporation
```
Comment (shermanyin) [inline modules/gcp/cac-igm/cac-cam.py.tmpl:1]: Thils file should not need to be a template

File: modules/gcp/cac-igm/main.tf (hunk 2 above)
Comment (shermanyin) [reply modules/gcp/cac-igm/main.tf:55]: Actually I don't see where the variables are used in the tmpl

File: deployments/gcp/multi-region/terraform.tfvars.sample (hunk 1 above)
Comment (Edwin-Pau) [reply deployments/gcp/multi-region/terraform.tfvars.sample:75]: Will discard this change - I believe I accidentally added this file in this branch. I've updated all sample files in the other branch (gcp-kms-script).

File: modules/gcp/cac-igm/cac-cam.py.tmpl (hunk 12 above)
Comment (Edwin-Pau) [reply modules/gcp/cac-igm/cac-cam.py.tmpl:84]: Looks like we don't need it - changed to read directly from CAM JSON.

**Comment by Edwin-Pau** [comment]
All comments addressed, will be creating new pull request with all the changes.

**Pull request closed without merging**